File: cpp/2539.count-the-number-of-good-subsequences/solution.cpp
// Created by none at 2025/05/26 13:11
// leetgo: dev
// https://leetcode.com/problems/count-the-number-of-good-subsequences/

#include "LC_IO.h"
#include <bits/stdc++.h>
#include <print>
#include <unordered_set>
#include <vector>
using namespace std;

// @lc code=begin

int pow(int a, int b, int n) {
  long long x = 1, y = a;
  while (b > 0) {
    if (b % 2 == 1) {
      x = (x * y) % n; // multiplying with base
    }
    y = (y * y) % n; // squaring the base
    b /= 2;
  }
  return x % n;
}

const int MOD = 1e9 + 7;

class Solution {
public:
  int countGoodSubsequences(string s) {
    vector<int> cnt(26, 0);
    int max = 0;
    for (auto &ch : s) {
      cnt[ch - 'a']++;
      max = std::max(max, cnt[ch - 'a']);
    }

    int res = 0;
    vector<long long> f(max + 1, 1);
    unordered_set<int> seen;
    for (int n = 0; n < max; n++) {
      long long cur = 1;
      seen.clear();
      for (auto m : cnt) {
        if (m > n) {
          if (!seen.contains(m)) {
            // comb(m,n+1) = comb(m,n) * (m-n) / (n+1)
            f[m] = f[m] * (m - n) % MOD * pow(n + 1, MOD - 2, MOD) % MOD;
            seen.insert(m);
          }
          cur = cur * (f[m] + 1) % MOD; // comb(m, n)
        }
      }
      res = (res + cur - 1) % MOD;
    }
    return res;
  }
};

// @lc code=end

int main() {
  ios_base::sync_with_stdio(false);
  stringstream out_stream;

  string s;
  LeetCodeIO::scan(cin, s);

  Solution *obj = new Solution();
  auto res = obj->countGoodSubsequences(s);
  LeetCodeIO::print(out_stream, res);
  cout << "\noutput: " << out_stream.rdbuf() << endl;

  delete obj;
  return 0;
}

File: python/2539.count-the-number-of-good-subsequences/solution.py
# Created by none at 2025/05/26 13:27
# leetgo: dev
# https://leetcode.com/problems/count-the-number-of-good-subsequences/

from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, cmp_to_key, lru_cache, reduce
from heapq import heapify, heappop, heappush, heappushpop, heapreplace
from itertools import accumulate, chain, count, pairwise, zip_longest
from math import ceil, comb, floor, gcd, inf, isqrt, lcm, log2, perm, sqrt
from operator import xor
from pprint import pprint
from string import ascii_lowercase
from typing import List, Optional

from leetgo_py import *

# @lc code=begin
mod = 10**9 + 7

N = 10**4 + 1
fac = [1] * N
for i in range(1, N):
    fac[i] = i * fac[i - 1] % mod


@cache
def my_comb(m: int, n: int):
    return fac[m] * pow(fac[m - n], -1, mod) % mod * pow(fac[n], -1, mod) % mod


class Solution:
    def countGoodSubsequences(self, s: str) -> int:
        cnt = Counter(s)
        res = 0
        for n in range(1, max(cnt.values()) + 1):
            cur = 1
            for m in cnt.values():
                if m >= n:
                    cur = cur * (my_comb(m, n) + 1) % mod
            res = (res + cur - 1) % mod
        return res


# @lc code=end

if __name__ == "__main__":
    s: str = deserialize("str", read_line())
    ans = Solution().countGoodSubsequences(s)
    print("\noutput:", serialize(ans, "integer"))
